storage/grpc_client.go

```diff
@@ -1470,18 +1487,32 @@ func (mr *gRPCBidiReader) wait() {
 
 // Close will notify stream manager goroutine that the reader has been closed, if it's still running.
 func (mr *gRPCBidiReader) close() error {
-	if mr.cancel != nil {
-		mr.cancel()
-	}
+	mr.closeManager <- true
+	mr.closeReceiver <- true
```

Comment:
There seem to be buffered channels. So it's not guaranteed that the receivers would have read the message when execution resumes. Even if thee were unbuffered channels, the sender would get unblocked as soon as the receiver gets the message. The sender and receiver would race to lock the mutex and the sender could win. This could result in the context getting cancelled before the stream is drained.